File: 71-simplify-path/simplify-path.java
class Solution {
    public String simplifyPath(String path) {
        Deque<String> stack = new ArrayDeque<>();
        String[] subDir = path.split("/+");

        for(String dir : subDir) {
            if (dir.isEmpty() || dir.equals(".")) {
                continue;
            } else if (dir.equals("..")){
                if (!stack.isEmpty()) {
                    stack.pop();
                }
            } else {
                stack.push(dir);
            }
        }

        StringBuilder ans = new StringBuilder();
        while(!stack.isEmpty()) {
            ans.insert(0, "/" + stack.pop());
        }

        return ans.length() == 0 ? "/" : ans.toString();
    }
}

File: 71-simplify-path/simplify-path.go
func simplifyPath(path string) string {
    dirs := make([]string, 0)
    subDir := strings.Split(path, "/")

    for _, dir := range subDir {
        if dir == "" || dir == "." {
            continue
        } else if dir == ".." {
            if len(dirs) > 0 {
                dirs = dirs[:len(dirs)-1]
            }
        } else {
            dirs = append(dirs, dir)
        }
    }

    ans := ""
    for _, d := range dirs {
        ans += "/" + d
    }

    if len(ans) == 0 {
        return "/"
    }
    return ans
}